### PR TITLE
Add IndexingParametersExtensions as a sample

### LINKED ISSUE #12154: Add IndexingParametersExtensions as a sample
Add this for track 2: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/search/Microsoft.Azure.Search.Service/src/Customizations/Indexers/Models/IndexingParametersExtensions.cs

### PR DESCRIPTION
**Pull request by heaths**
Fixes #12154

### CODE REVIEW

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:13]: Why isn't this in the swagger instead?

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs:14]: Why are these extension methods?

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs:135]: Should we take a separator type or format (that's only CSV today)?  The params gives me a bit of caution in case it's going to be a string.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParameters.cs
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParameters.cs:13]: This design feels odd.  Why not just a bunch of property getters/setters that show people all the known options that use this Configuration as a backing store?

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:31]: We should call this `Metadata` to match Track 2 Storage names

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:37]: This should be `MetadataAndProperties`

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:40]: Does that mean Metadata + Properties when you say all metadata?

**Review by tg-msft — APPROVED** [review]
I'm approving if you want to get this merged and iterate, but I don't really understand the design.  Do you have any insight into why Track 1 did it this way?

**Comment by tg-msft** [comment]
Actually, I might not understand the different ExtractionMode values and what they mean.  Maybe we should be careful about the word Metadata since it's heavily overloaded across these contexts?

**Comment by brjohnstmsft** [comment]
@tg-msft I can shed some light on the design here. Way back in 2014 when we first introduced indexers, the API surface was very unstable, especially the configuration parameters for indexer execution. The decision was made to model indexing parameters as an untyped property bag since the number of parameters grew very quickly. We needed a way to let customers use newly-defined parameters even between SDK releases, which is why it's just a dictionary with some syntactic sugar on top.

Now that some time has passed, perhaps we should consider making the stable indexing parameters strongly-typed by adding them to the Swagger. That's a call for @bleroy and team to make. However, it's likely too late to do that for GA. Still, adding strongly-typed properties later should be a non-breaking change, right?

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs:135]: That's the format that the service expects for this property in the bag. I'll defer to @bleroy and @jennifermarsman if it would be sensible to allow the user to define others for possible future use.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:31]: I'd prefer the "enum" names match the actual value, which is what it would have to be on JS and Python enums as well. I believe Java for extensible enums has the same design.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:40]: Same as above: extensible enum names should match enum values.

**Comment by heaths** [comment]
Yes, we can add properties later but not deprecate the extension methods, though we can hide them for future use. That said, if this design is of concern it may be best to leave off for now, though it leaves a bit of a gap with track one.

/cc @bleroy @jennifermarsman

**Comment by heaths** [comment]
After much discussion, we're instead going to turn this into a sample for now that devs can copy and paste, along with sample docs how to use it until we get changes into swagger to add this for all languages as proper model properties. We want to align "metadata" with how Storage refers to it, so we need to find out exactly what the behavior of the service is so we can appropriately (re)name the properties.

Given that the model properties will be used as-is, I'll rename `IncludeFileNameExtensions` back to `IndexFileNameExtensions`, though if customers are using these and eventual get the model updates in a new version of Search, we might want to make the method names verb-form like `SetIndexFileNameExtensions`, etc. See how the compiler reacts to extension methods and property names having the same name (declared methods always override extension methods with the same parameter types, but not sure about property/extension method name collisions).

/cc @brjohnstmsft @jennifermarsman @bleroy @tg-msft @AlexGhiondea

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs
Comment (jennifermarsman) [inline sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:528]: Nit: Could we make this ExcludedFileNameExtensions?

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs
Comment (jennifermarsman) [inline sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:529]: These are called "IndexedExtensions" on the portal.  Would be nice to have consistency.  At minimum, I prefer "IncludedFileNameExtensions" to "IncludeFileNameExtensions" unless there is huge precedent elsewhere.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs
Comment (jennifermarsman) [inline sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:530]: There is also a "Default" parse mode available on the Azure portal besides the other five you enumerated (Text, Json, JsonArrays, JsonLines, DelimitedText).  Are we exposing that?

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (jennifermarsman) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:13]: +1: this seems like it should be in the swagger

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (jennifermarsman) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:31]: I agree that I would like those names to stay the same, for consistency with the user experience on the Azure portal, which gives choices of ContentAndMetadata, AllMetadata, and StorageMetadata.  (We can change portal if we must but consistency in naming is important.)

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (jennifermarsman) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:37]: Why "MetadataAndProperties"?  If we change here, I want it changed in portal as well.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (jennifermarsman) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:40]: Sounds like we should have a naming conversation and make it consistent everywhere.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs
Comment (jennifermarsman) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs:135]: In the portal, we have an open text box to accept a string for the delimiter character (it does default to comma).  So yes, I do think we need to add functionality to accept that here.  

There is also a boolean (checkbox in the portal) to specify whether the first line contains headers.  If it is unchecked, the headers open text box appears.  The below logic looks correct.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (tg-msft) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:37]: Storage makes a distinction between [metadata](https://docs.microsoft.com/en-us/rest/api/storageservices/get-blob-metadata) and [properties](https://docs.microsoft.com/en-us/rest/api/storageservices/get-blob-properties). 
 Metadata are custom key/value pairs you can throw on a blob.  Properties are well known HTTP headers like Content-Type.  (FWIW, this is a newer distinction in the last couple of years.)  As somebody familiar with Storage, I'm not sure what `AllMetadata` means here.  Does it include both Properties and Metadata?

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:528]: Since we're going to likely be defining these in swagger and will generate the properties as-is on models, I'd rather the eventual property names match the JSON schema property names. So, as samples, I'll probably be changing these related extension methods to `Set<would-be model property names>` so we don't conflict later if someone is using these samples in their code.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:530]: I can. What does it do? Note, all this code was ported from track 1 that @brjohnstmsft worked on. The only thing I changed (apart from some implementation details - see above), was the `IndexFileNameExtensions` to `IncludeFileNameExstensions`, which I'm going to rename back.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs:14]: As discussed offline, we'll implement these as a sample for now until we can make these proper properties in the swagger post-GA.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/BlobExtractionMode.cs:37]: For the sample, I'm keeping the names as they show in the portal. When we code these into the swagger, we should decide example what they should specify (not breaking existing customers) and how they should display.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersExtensions.cs:14]: We'll make them proper properties post-GA.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs]: Seems we don't need to since it is the default, but let me know if that's an incorrect assumption.

**Review by bleroy — APPROVED** [review]

**Pull request merged**
